tephra-hbase-compat-2.0/src/main/java/org/apache/tephra/hbase/coprocessor/FilteredInternalScanner.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.tephra.hbase.coprocessor;
+
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.filter.Filter;
+import org.apache.hadoop.hbase.filter.Filter.ReturnCode;
+import org.apache.hadoop.hbase.regionserver.InternalScanner;
+import org.apache.hadoop.hbase.regionserver.ScannerContext;
+import org.apache.tephra.hbase.coprocessor.TransactionProcessor.IncludeInProgressFilter;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Wrapper of InternalScanner to apply Transaction visibility filter for flush and compact
+ */
+public class FilteredInternalScanner implements InternalScanner {
+
+  private InternalScanner delegate;
+  private Filter filter;
+
+  public FilteredInternalScanner(InternalScanner internalScanner, IncludeInProgressFilter filter) {
+    this.delegate = internalScanner;
+    this.filter = filter;
+  }
+
+  @Override
+  public void close() throws IOException {
+    this.delegate.close();
+  }
+
+  @Override
+  public boolean next(List<Cell> result, ScannerContext scannerContext) throws IOException {
+    List<Cell> outResult = new ArrayList<Cell>();
+    while (true) {
+      boolean next = delegate.next(outResult, scannerContext);
+      for (Cell cell : outResult) {
+        ReturnCode code = filter.filterKeyValue(cell);
+        switch (code) {
+        // included, so we are done
+        case INCLUDE:
+        case INCLUDE_AND_NEXT_COL:
```

Comment:
Yes, so on next cell, we expect the filter to return some NEXT_COL/SKIP (if it returned INCLUDE_AND_NEXT_COL for the previous cell) so that we will skip them.

Reply:
👍  Just confirmed that transaction filter does this

Reply:
The only downside I see with this approach is that we'll iterate through all the cells for a column, even though we could have skipped some cells when the filter returns `INCLUDE_AND_NEXT_COL`. 

For example if there are 100 versions of a value in a column, and the filter figures out at the 10th version that the rest of the versions have reached the TTL, then in the earlier code we would not  iterate through the rest of the 90 cells (or at least that's what I think was happening in the underlying scanner). Is there anyway we can achieve the same result with the new APIs?

Reply:
Yes @poornachandra , that's a downside, but with HBase-2.0 we don't have access to create StoreScanner with the custom scan in pre-compact/pre-flush hooks. 
I had a discussion with Anoop(From HBase) on https://issues.apache.org/jira/browse/TEPHRA-274 for reviewing the approach, probably currently there is no other way.